service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/dataproccluster/UpdateDataprocClusterStep.java

```diff
@@ -0,0 +1,149 @@
+import com.google.api.client.googleapis.json.GoogleJsonResponseException;
+import com.google.api.services.dataproc.model.AutoscalingConfig;
+import com.google.api.services.dataproc.model.Cluster;
+import com.google.api.services.dataproc.model.ClusterConfig;
+import com.google.api.services.dataproc.model.InstanceGroupConfig;
+import com.google.api.services.dataproc.model.LifecycleConfig;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import org.springframework.http.HttpStatus;
+
+public class UpdateDataprocClusterStep implements Step {
+
+  private final ControlledDataprocClusterResource resource;
+  private final CrlService crlService;
+
+  // Mask paths for cluster config fields that are updatable
+  private static final String NUM_WORKERS = "config.worker_config.num_instances";
+  private static final String NUM_SECONDARY_WORKERS =
+      "config.secondary_worker_config.num_instances";
+  private static final String AUTOSCALING_POLICY = "config.autoscaling_config.policy_uri";
+  // Only clusters created with scheduled deletion enabled can be updated. Removing scheduled
+  // deletion from a cluster, i.e. settling lifecycleConfig to be empty is irreversible.
+  private static final String IDLE_DELETE_TTL = "config.lifecycle_config.idle_delete_ttl";
+  private static final String AUTO_DELETE_TTL = "config.lifecycle_config.auto_delete_ttl";
+  private static final String AUTO_DELETE_TIME = "config.lifecycle_config.auto_delete_time";
+
+  public UpdateDataprocClusterStep(
+      ControlledDataprocClusterResource resource, CrlService crlService) {
+    this.resource = resource;
+    this.crlService = crlService;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    ApiControlledDataprocClusterUpdateParameters updateParameters =
+        flightContext
+            .getInputParameters()
+            .get(UPDATE_PARAMETERS, ApiControlledDataprocClusterUpdateParameters.class);
+    if (updateParameters == null) {
+      return StepResult.getStepResultSuccess();
+    }
+    return updateCluster(updateParameters);
+  }
+
+  // Restore the previous values of the update parameters
+  @Override
+  public StepResult undoStep(FlightContext flightContext) throws InterruptedException {
+    FlightMap workingMap = flightContext.getWorkingMap();
+    ApiControlledDataprocClusterUpdateParameters prevParameters =
+        FlightUtils.getRequired(
+            workingMap,
+            PREVIOUS_UPDATE_PARAMETERS,
+            ApiControlledDataprocClusterUpdateParameters.class);
+    return updateCluster(prevParameters);
+  }
+
+  // Update the cluster with the given update parameters.
+  private StepResult updateCluster(ApiControlledDataprocClusterUpdateParameters updateParameters) {
+    ClusterName clusterName = resource.toClusterName();
+    List<String> updateMaskPaths = new ArrayList<>();
+    Cluster updateClusterMask =
```

Comment:
This name is a bit confusing - iiuc this isn't a "mask", this is the updated cluster. Is that right?

Reply:
I named it `updateClusterMask` since only the relevant updated values are set in the object while all other fields are empty. `updateClusterValues` or `valuesToUpdate` as alternatives?

Reply:
`updatedCluster` and `updateMaskPaths` make sense to me as the cluster and mask, thanks!